osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineControlPointDisplay.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Extensions.IEnumerableExtensions;
+using osu.Framework.Graphics;
+using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.ControlPoints;
+using osu.Game.Graphics;
+using osu.Game.Screens.Edit.Components.Timelines.Summary.Parts;
+using osu.Game.Screens.Edit.Components.Timelines.Summary.Visualisations;
+
+namespace osu.Game.Screens.Edit.Compose.Components.Timeline
+{
+    public class TimelineControlPointDisplay : TimelinePart
+    {
+        protected override void LoadBeatmap(WorkingBeatmap beatmap)
+        {
+            base.LoadBeatmap(beatmap);
+
+            ControlPointInfo cpi = beatmap.Beatmap.ControlPointInfo;
+
+            cpi.TimingPoints.ForEach(addTimingPoint);
+
+            // Consider all non-timing points as the same type
+            cpi.SamplePoints.Select(c => (ControlPoint)c)
+               .Concat(cpi.EffectPoints)
+               .Concat(cpi.DifficultyPoints)
+               .Distinct()
+               // Non-timing points should not be added where there are timing points
+               .Where(c => cpi.TimingPointAt(c.Time).Time != c.Time)
```

Comment:
This looks like a potentially expensive quadratic computation. Possibly worth to have a look if this can be optimised further.